### ISSUE TITLE
SpiderNode w/external SpiderMonkey crashes in REPL with "compartment mismatch" error

### ISSUE THREAD

**Issue by mykmelez** [mozilla/spidernode#196]
After #193 and #195, SpiderNode compiles and links with an external SpiderMonkey, and it runs and evaluates script specified by the _-e_ flag correctly:

``` bash
> ./node -e 'console.log("hello from " + process.jsEngine)'
hello from spidermonkey
```

But evaluating the same script in the REPL crashes with a "compartment mismatch" error:

``` bash
> ./node
> console.log("hello from " + process.jsEngine)
*** Compartment mismatch 0x104e2ac00 vs. 0x1040ba000
Segmentation fault: 11
```

**Comment by ehsan** [comment]
Can you get a backtrace please?

**Comment by brendandahl** [comment]
```
 bt
* thread #1: tid = 0xb8457, 0x0000000100e0295a node`js::CompartmentChecker::fail(c1=0x0000000105d1d000, c2=0x0000000104917800) + 90 at jscntxtinlines.h:40, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x0000000100e0295a node`js::CompartmentChecker::fail(c1=0x0000000105d1d000, c2=0x0000000104917800) + 90 at jscntxtinlines.h:40
    frame #1: 0x0000000100e028e2 node`js::CompartmentChecker::check(this=0x00007fff5fbd6d10, c=0x0000000104917800) + 130 at jscntxtinlines.h:61
    frame #2: 0x0000000100e027d9 node`js::CompartmentChecker::check(this=0x00007fff5fbd6d10, obj=0x0000000105c72560) + 217 at jscntxtinlines.h:73
    frame #3: 0x0000000100af6323 node`js::CompartmentChecker::check(this=0x00007fff5fbd6d10, v=0x00007fff5fbd6e08) + 67 at jscntxtinlines.h:94
    frame #4: 0x0000000100af6190 node`void js::CompartmentChecker::check<JS::Value>(this=0x00007fff5fbd6d10, handle=Handle<JS::Value> @ 0x00007fff5fbd6ce8) + 48 at jscntxtinlines.h:83
    frame #5: 0x0000000100ae395a node`void js::assertSameCompartment<JS::Handle<JS::Value>, JS::Handle<JS::Value> >(cx=0x0000000104948000, t1=0x00007fff5fbd6d88, t2=0x00007fff5fbd6d80) + 122 at jscntxtinlines.h:180
    frame #6: 0x0000000100aadb1e node`JS_LooselyEqual(cx=0x0000000104948000, value1=JS::HandleValue @ 0x00007fff5fbd6d88, value2=JS::HandleValue @ 0x00007fff5fbd6d80, equal=0x00007fff5fbd6ddf) + 110 at jsapi.cpp:405
    frame #7: 0x000000010031a4fe node`v8::Value::Equals(this=0x0000000105fa22b0, context=Local<v8::Context> @ 0x00007fff5fbd6e50, that=v8::Handle<v8::Value> @ 0x00007fff5fbd6e48) const + 270 at v8value.cc:308
    frame #8: 0x000000010031a5d2 node`v8::Value::Equals(this=0x0000000105fa22b0, that=v8::Handle<v8::Value> @ 0x00007fff5fbd6ea8) const + 66 at v8value.cc:315
    frame #9: 0x0000000100263d15 node`bool v8::Local<v8::Value>::operator==<v8::Object>(this=0x00007fff5fbd6f18, that=0x00007fff5fbd6f58) const + 69 at v8.h:3286
    frame #10: 0x0000000100263560 node`node::ContextifyContext::GlobalPropertyGetterCallback(property=Local<v8::Name> @ 0x00007fff5fbd6f88, args=0x00007fff5fbd70a0) + 320 at node_contextify.cc:364
    frame #11: 0x000000010030130b node`(anonymous namespace)::PropCallbackTraits<void (isolate=0x00000001049040b8, callback=(node`node::ContextifyContext::GlobalPropertyGetterCallback(v8::Local<v8::Name>, v8::PropertyCallbackInfo<v8::Value> const&) at node_contextify.cc:345), id=JS::HandleId @ 0x00007fff5fbd6ff8, info=0x00007fff5fbd70a0, vp=JS::MutableHandleValue @ 0x00007fff5fbd6ff0)(v8::Local<v8::Name>, v8::PropertyCallbackInfo<v8::Value> const&), v8::Name>::doCall(v8::Isolate*, void (*)(v8::Local<v8::Name>, v8::PropertyCallbackInfo<v8::Value> const&), JS::Handle<jsid>, v8::PropertyCallbackInfo<v8::Value>&, JS::MutableHandle<JS::Value>) + 75 at v8objecttemplate.cc:357
    frame #12: 0x0000000100301dcf node`bool (anonymous namespace)::ResolveOpImpl_Getter<void (cx=0x0000000104948000, obj=JS::HandleObject @ 0x00007fff5fbd7180, id=JS::HandleId @ 0x00007fff5fbd7178, resolved=0x00007fff5fbd729b)(v8::Local<v8::Name>, v8::PropertyCallbackInfo<v8::Value> const&), v8::Name>(JSContext*, JS::Handle<JSObject*>, JS::Handle<jsid>, bool*) + 591 at v8objecttemplate.cc:597
    frame #13: 0x00000001002fceb0 node`(anonymous namespace)::ResolveOp_Getter(cx=0x0000000104948000, obj=JS::HandleObject @ 0x00007fff5fbd71f0, id=JS::HandleId @ 0x00007fff5fbd71e8, resolved=0x00007fff5fbd729b) + 432 at v8objecttemplate.cc:641
    frame #14: 0x0000000100bf8c90 node`js::CallResolveOp(cx=0x0000000104948000, obj=js::HandleNativeObject @ 0x00007fff5fbd7310, id=JS::HandleId @ 0x00007fff5fbd7308, propp=js::MutableHandleShape @ 0x00007fff5fbd7300, recursedp=0x00007fff5fbd738f) + 320 at NativeObject-inl.h:397
    frame #15: 0x0000000100bf8ab8 node`bool js::LookupOwnPropertyInline<(js::AllowGC)1>(cx=0x0000000104948000, obj=js::MaybeRooted<js::NativeObject *, js::AllowGC>::HandleType @ 0x00007fff5fbd73f0, id=js::MaybeRooted<jsid, js::AllowGC>::HandleType @ 0x00007fff5fbd73e8, propp=js::MaybeRooted<js::Shape *, js::AllowGC>::MutableHandleType @ 0x00007fff5fbd73e0, donep=0x00007fff5fbd74c7)1>::HandleType, js::MaybeRooted<jsid, (js::AllowGC)1>::HandleType, js::MaybeRooted<js::Shape*, (js::AllowGC)1>::MutableHandleType, bool*) + 680 at NativeObject-inl.h:485
    frame #16: 0x0000000100bd4db4 node`bool js::LookupPropertyInline<(js::AllowGC)1>(cx=0x0000000104948000, obj=js::MaybeRooted<js::NativeObject *, js::AllowGC>::HandleType @ 0x00007fff5fbd7500, id=js::MaybeRooted<jsid, js::AllowGC>::HandleType @ 0x00007fff5fbd74f8, objp=js::MaybeRooted<JSObject *, js::AllowGC>::MutableHandleType @ 0x00007fff5fbd74f0, propp=js::MaybeRooted<js::Shape *, js::AllowGC>::MutableHandleType @ 0x00007fff5fbd74e8)1>::HandleType, js::MaybeRooted<jsid, (js::AllowGC)1>::HandleType, js::MaybeRooted<JSObject*, (js::AllowGC)1>::MutableHandleType, js::MaybeRooted<js::Shape*, (js::AllowGC)1>::MutableHandleType) + 132 at NativeObject-inl.h:559
    frame #17: 0x0000000100bd4d0f node`js::LookupProperty(cx=0x0000000104948000, obj=JS::HandleObject @ 0x00007fff5fbd7590, id=JS::HandleId @ 0x00007fff5fbd7588, objp=JS::MutableHandleObject @ 0x00007fff5fbd7580, propp=js::MutableHandleShape @ 0x00007fff5fbd7578) + 207 at jsobj.cpp:2147
    frame #18: 0x0000000100bd5123 node`js::LookupName(cx=0x0000000104948000, name=js::HandlePropertyName @ 0x00007fff5fbd7660, envChain=JS::HandleObject @ 0x00007fff5fbd7658, objp=JS::MutableHandleObject @ 0x00007fff5fbd7650, pobjp=JS::MutableHandleObject @ 0x00007fff5fbd7648, propp=js::MutableHandleShape @ 0x00007fff5fbd7640) + 227 at jsobj.cpp:2157
    frame #19: 0x0000000100e4f8fd node`GetNameOperation(cx=0x0000000104948000, fp=0x0000000104ea5978, pc="\x9a", vp=JS::MutableHandleValue @ 0x00007fff5fbd77e0) + 509 at Interpreter.cpp:223
    frame #20: 0x0000000100e213c3 node`Interpret(cx=0x0000000104948000, state=0x00007fff5fbdaa60) + 47459 at Interpreter.cpp:3057
    frame #21: 0x0000000100e15972 node`js::RunScript(cx=0x0000000104948000, state=0x00007fff5fbdaa60) + 770 at Interpreter.cpp:400
    frame #22: 0x0000000100e2c9b3 node`js::ExecuteKernel(cx=0x0000000104948000, script=JS::HandleScript @ 0x00007fff5fbdab20, envChainArg=0x0000000105c79040, newTargetValue=0x00007fff5fbdab70, evalInFrame=(ptr_ = 0), result=0x00007fff5fbdade8) + 755 at Interpreter.cpp:681
    frame #23: 0x0000000100e2cd13 node`js::Execute(cx=0x0000000104948000, script=JS::HandleScript @ 0x00007fff5fbdabc8, envChainArg=0x0000000105c79040, rval=0x00007fff5fbdade8) + 771 at Interpreter.cpp:713
    frame #24: 0x0000000100ac1a90 node`ExecuteScript(cx=0x0000000104948000, scope=JS::HandleObject @ 0x00007fff5fbdac38, script=JS::HandleScript @ 0x00007fff5fbdac30, rval=0x00007fff5fbdade8) + 416 at jsapi.cpp:4288
    frame #25: 0x0000000100ac1fb1 node`JS::CloneAndExecuteScript(cx=0x0000000104948000, scriptArg=JS::HandleScript @ 0x00007fff5fbdad00, rval=JS::MutableHandleValue @ 0x00007fff5fbdacf8) + 401 at jsapi.cpp:4351
    frame #26: 0x000000010030d08c node`v8::Script::Run(this=0x000000010594d870, context=Local<v8::Context> @ 0x00007fff5fbdae00) + 364 at v8script.cc:110
    frame #27: 0x000000010030d1f2 node`v8::Script::Run(this=0x000000010594d870) + 50 at v8script.cc:120
    frame #28: 0x0000000100266228 node`node::ContextifyScript::EvalMachine(env=0x00007fff5fbfed60, timeout=-1, display_errors=false, break_on_sigint=true, args=0x00007fff5fbdb3b0, try_catch=0x00007fff5fbdb1e8) + 488 at node_contextify.cc:849
    frame #29: 0x0000000100264a08 node`node::ContextifyScript::RunInContext(args=0x00007fff5fbdb3b0) + 504 at node_contextify.cc:609
    frame #30: 0x00000001002d8ca2 node`v8::internal::FunctionCallback::NativeFunctionCallback(cx=0x0000000104948000, argc=2, vp=0x00007fff5fbdb8a8) + 1682 at v8function.cc:185
    frame #31: 0x0000000100e57c7d node`js::CallJSNative(cx=0x0000000104948000, native=(node`v8::internal::FunctionCallback::NativeFunctionCallback(JSContext*, unsigned int, JS::Value*) at v8function.cc:117), args=0x00007fff5fbdb850)(JSContext*, unsigned int, JS::Value*), JS::CallArgs const&) + 173 at jscntxtinlines.h:235
    frame #32: 0x0000000100e2b5a4 node`js::InternalCallOrConstruct(cx=0x0000000104948000, args=0x00007fff5fbdb850, construct=NO_CONSTRUCT) + 1028 at Interpreter.cpp:454
    frame #33: 0x0000000100e2ba3d node`InternalCall(cx=0x0000000104948000, args=0x00007fff5fbdb850) + 493 at Interpreter.cpp:499
    frame #34: 0x0000000100e2bac6 node`js::Call(cx=0x0000000104948000, fval=JS::HandleValue @ 0x00007fff5fbdb790, thisv=JS::HandleValue @ 0x00007fff5fbdb788, args=0x00007fff5fbdb850, rval=JS::MutableHandleValue @ 0x00007fff5fbdb780) + 118 at Interpreter.cpp:518
    frame #35: 0x0000000100b405ab node`js::fun_call(cx=0x0000000104948000, argc=3, vp=0x0000000104ea5948) + 587 at jsfun.cpp:1249
    frame #36: 0x0000000100e57c7d node`js::CallJSNative(cx=0x0000000104948000, native=(node`js::fun_call(JSContext*, unsigned int, JS::Value*) at jsfun.cpp:1218), args=0x00007fff5fbddc28)(JSContext*, unsigned int, JS::Value*), JS::CallArgs const&) + 173 at jscntxtinlines.h:235
    frame #37: 0x0000000100e2b5a4 node`js::InternalCallOrConstruct(cx=0x0000000104948000, args=0x00007fff5fbddc28, construct=NO_CONSTRUCT) + 1028 at Interpreter.cpp:454
    frame #38: 0x0000000100e2ba3d node`InternalCall(cx=0x0000000104948000, args=0x00007fff5fbddc28) + 493 at Interpreter.cpp:499
    frame #39: 0x0000000100e2b83d node`js::CallFromStack(cx=0x0000000104948000, args=0x00007fff5fbddc28) + 29 at Interpreter.cpp:505
    frame #40: 0x0000000100e20415 node`Interpret(cx=0x0000000104948000, state=0x00007fff5fbded38) + 43445 at Interpreter.cpp:2915
    frame #41: 0x0000000100e15972 node`js::RunScript(cx=0x0000000104948000, state=0x00007fff5fbded38) + 770 at Interpreter.cpp:400
    frame #42: 0x0000000100e2b671 node`js::InternalCallOrConstruct(cx=0x0000000104948000, args=0x00007fff5fbe0fa8, construct=NO_CONSTRUCT) + 1233 at Interpreter.cpp:472
    frame #43: 0x0000000100e2ba3d node`InternalCall(cx=0x0000000104948000, args=0x00007fff5fbe0fa8) + 493 at Interpreter.cpp:499
    frame #44: 0x0000000100e2b83d node`js::CallFromStack(cx=0x0000000104948000, args=0x00007fff5fbe0fa8) + 29 at Interpreter.cpp:505
    frame #45: 0x0000000100e20415 node`Interpret(cx=0x0000000104948000, state=0x00007fff5fbe20b8) + 43445 at Interpreter.cpp:2915
    frame #46: 0x0000000100e15972 node`js::RunScript(cx=0x0000000104948000, state=0x00007fff5fbe20b8) + 770 at Interpreter.cpp:400
    frame #47: 0x0000000100e2b671 node`js::InternalCallOrConstruct(cx=0x0000000104948000, args=0x00007fff5fbe27a0, construct=NO_CONSTRUCT) + 1233 at Interpreter.cpp:472
    frame #48: 0x0000000100e2ba3d node`InternalCall(cx=0x0000000104948000, args=0x00007fff5fbe27a0) + 493 at Interpreter.cpp:499
    frame #49: 0x0000000100e2bac6 node`js::Call(cx=0x0000000104948000, fval=JS::HandleValue @ 0x00007fff5fbe2210, thisv=JS::HandleValue @ 0x00007fff5fbe2208, args=0x00007fff5fbe27a0, rval=JS::MutableHandleValue @ 0x00007fff5fbe2200) + 118 at Interpreter.cpp:518
    frame #50: 0x0000000100b410b3 node`js::fun_apply(cx=0x0000000104948000, argc=2, vp=0x0000000104ea5698) + 1427 at jsfun.cpp:1318
    frame #51: 0x0000000100e57c7d node`js::CallJSNative(cx=0x0000000104948000, native=(node`js::fun_apply(JSContext*, unsigned int, JS::Value*) at jsfun.cpp:1255), args=0x00007fff5fbe4b78)(JSContext*, unsigned int, JS::Value*), JS::CallArgs const&) + 173 at jscntxtinlines.h:235
    frame #52: 0x0000000100e2b5a4 node`js::InternalCallOrConstruct(cx=0x0000000104948000, args=0x00007fff5fbe4b78, construct=NO_CONSTRUCT) + 1028 at Interpreter.cpp:454
    frame #53: 0x0000000100e2ba3d node`InternalCall(cx=0x0000000104948000, args=0x00007fff5fbe4b78) + 493 at Interpreter.cpp:499
    frame #54: 0x0000000100e2b83d node`js::CallFromStack(cx=0x0000000104948000, args=0x00007fff5fbe4b78) + 29 at Interpreter.cpp:505
    frame #55: 0x0000000100e20415 node`Interpret(cx=0x0000000104948000, state=0x00007fff5fbe5c88) + 43445 at Interpreter.cpp:2915
    frame #56: 0x0000000100e15972 node`js::RunScript(cx=0x0000000104948000, state=0x00007fff5fbe5c88) + 770 at Interpreter.cpp:400
    frame #57: 0x0000000100e2b671 node`js::InternalCallOrConstruct(cx=0x0000000104948000, args=0x00007fff5fbe5ea0, construct=NO_CONSTRUCT) + 1233 at Interpreter.cpp:472
    frame #58: 0x0000000100e2ba3d node`InternalCall(cx=0x0000000104948000, args=0x00007fff5fbe5ea0) + 493 at Interpreter.cpp:499
    frame #59: 0x0000000100e2bac6 node`js::Call(cx=0x0000000104948000, fval=JS::HandleValue @ 0x00007fff5fbe5de0, thisv=JS::HandleValue @ 0x00007fff5fbe5dd8, args=0x00007fff5fbe5ea0, rval=JS::MutableHandleValue @ 0x00007fff5fbe5dd0) + 118 at Interpreter.cpp:518
    frame #60: 0x0000000100b405ab node`js::fun_call(cx=0x0000000104948000, argc=2, vp=0x0000000104ea5498) + 587 at jsfun.cpp:1249
    frame #61: 0x0000000100e57c7d node`js::CallJSNative(cx=0x0000000104948000, native=(node`js::fun_call(JSContext*, unsigned int, JS::Value*) at jsfun.cpp:1218), args=0x00007fff5fbe8278)(JSContext*, unsigned int, JS::Value*), JS::CallArgs const&) + 173 at jscntxtinlines.h:235
    frame #62: 0x0000000100e2b5a4 node`js::InternalCallOrConstruct(cx=0x0000000104948000, args=0x00007fff5fbe8278, construct=NO_CONSTRUCT) + 1028 at Interpreter.cpp:454
    frame #63: 0x0000000100e2ba3d node`InternalCall(cx=0x0000000104948000, args=0x00007fff5fbe8278) + 493 at Interpreter.cpp:499
    frame #64: 0x0000000100e2b83d node`js::CallFromStack(cx=0x0000000104948000, args=0x00007fff5fbe8278) + 29 at Interpreter.cpp:505
    frame #65: 0x0000000100e20415 node`Interpret(cx=0x0000000104948000, state=0x00007fff5fbe9388) + 43445 at Interpreter.cpp:2915
    frame #66: 0x0000000100e15972 node`js::RunScript(cx=0x0000000104948000, state=0x00007fff5fbe9388) + 770 at Interpreter.cpp:400
    frame #67: 0x0000000100e2b671 node`js::InternalCallOrConstruct(cx=0x0000000104948000, args=0x00007fff5fbe96b0, construct=NO_CONSTRUCT) + 1233 at Interpreter.cpp:472
    frame #68: 0x0000000100e2ba3d node`InternalCall(cx=0x0000000104948000, args=0x00007fff5fbe96b0) + 493 at Interpreter.cpp:499
    frame #69: 0x0000000100e2b83d node`js::CallFromStack(cx=0x0000000104948000, args=0x00007fff5fbe96b0) + 29 at Interpreter.cpp:505
    frame #70: 0x00000001005da30f node`js::jit::DoCallFallback(cx=0x0000000104948000, frame=0x00007fff5fbe9948, stub_=0x000000010542cd50, argc=4, vp=0x00007fff5fbe9860, res=JS::MutableHandleValue @ 0x00007fff5fbe97b8) + 1999 at BaselineIC.cpp:5993
    frame #71: 0x0000000104a1f461
```

**Comment by ehsan** [comment]
Sweet!  I fixed it this morning in 93c5b086ea6bd69582ac606895b6aa6094ddf618.  :-)
